### PR TITLE
T5241: Support netns for veth and dummy interfaces (second edition)

### PR DESCRIPTION
**Pull request by sever-sever**
## Change Summary

The second version of netns implementation
The first version is there https://github.com/vyos/vyos-1x/pull/2020

It changes  `_cmd` for `conftrol.py` to use correct network namespace

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Relate PR
* https://github.com/vyos/vyos-build/pull/354

## Related Task(s)

* https://vyos.dev/T5241

## Component(s) name

netns
## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: python/vyos/ifconfig/interface.py [hunk 1]
```diff
@@ -59,6 +59,36 @@
...
 
+
+def _interface_exists_in_netns(interface_name, netns):
+    from vyos.util import rc_cmd
+    rc, out = rc_cmd(f'ip netns exec {netns} ip link show dev {interface_name}')
+    if rc == 0:
+        return True
+    return False
+
+def _is_intf_addr_assigned(ifname, addr, netns=None):
```
Comment (c-po) [inline python/vyos/ifconfig/interface.py]: Why not use the `vyos.validate.is_intf_addr_assigned()` function end add an optional netns parameter?

File: python/vyos/ifconfig/interface.py [hunk 2]
```diff
@@ -282,7 +309,7 @@ def get_config(cls):
         """
         return deepcopy(cls.default)
 
-    def __init__(self, ifname, **kargs):
+    def __init__(self, ifname, netns='', **kargs):
```
Comment (c-po) [inline python/vyos/ifconfig/interface.py]: Why is this needed? `netns` key is part of `**kargs` and will thus be available as `self.config['netns']`

File: python/vyos/ifconfig/interface.py [hunk 3]
```diff
@@ -339,7 +367,12 @@ def __init__(self, ifname, **kargs):
         self.vrrp = VRRP(ifname)
 
     def _create(self):
-        cmd = 'ip link add dev {ifname} type {type}'.format(**self.config)
+        # Do not create interface that already exist or exists in netns
+        #if self.exists(f'{self.ifname}') or get_interface_namespace(self.config['ifname']):
+        if self.exists(f'{self.ifname}') or _interface_exists_in_netns(self.config['ifname'], self.netns):
```
Comment (c-po) [inline python/vyos/ifconfig/interface.py]: Why not extend `vyos.ifconfig.interface.exists()` and make it netns aware with an optional argument? So it would check if the interface exists either in the global context or the netns context - for any match it will return True

File: python/vyos/ifconfig/interface.py [hunk 4]
```diff
@@ -339,7 +367,12 @@ def __init__(self, ifname, **kargs):
...
+            return
+        cmd = f'ip netns exec {self.netns} ' if self.netns else ''
+        cmd += 'ip link add dev {ifname} type {type}'.format(**self.config)
```
Comment (c-po) [inline python/vyos/ifconfig/interface.py]: I'd write it the other way arround, add the specialty later on

```python
cmd = 'ip link add dev {ifname} type {type}'.format(**self.config)
if 'netns' in self.config: cmd = f'ip netns exec {self.config["netns"]} {cmd}'
```

File: python/vyos/ifconfig/interface.py [hunk 5]
```diff
@@ -373,14 +406,21 @@ def _delete(self):
         # NOTE (Improvement):
         # after interface removal no other commands should be allowed
         # to be called and instead should raise an Exception:
-        cmd = 'ip link del dev {ifname}'.format(**self.config)
+        netns = get_interface_namespace(self.config['ifname'])
```
Comment (c-po) [inline python/vyos/ifconfig/interface.py:393]: Why not use the netns name from vyos CLI config by accessing self.config()?

File: python/vyos/ifconfig/interface.py [hunk 6]
```diff
@@ -739,6 +781,11 @@ def set_ipv4_source_validation(self, value):
 
         As per RFC3074.
         """
+
+        # Don't allow for netns yet
+        if self.netns:
```
Comment (c-po) [inline python/vyos/ifconfig/interface.py]: instead of silently hiding this we should use common verify() rules and inform the user

**Review by c-po — CHANGES_REQUESTED** [review]

File: python/vyos/ifconfig/interface.py (hunk 2 above)
Comment (sever-sever) [reply python/vyos/ifconfig/interface.py]: It also used for  parent class “control.py” as self.netns
Should we somehow get **kargs in the parent class?

Comment (c-po) [reply python/vyos/ifconfig/interface.py]: self.config is accessible in the base class, too

File: python/vyos/ifconfig/interface.py (hunk 1 above)
Comment (sever-sever) [reply python/vyos/ifconfig/interface.py]: The current original module uses `netifaces`, `ifaddresses` and `AF_INET` https://github.com/vyos/vyos-1x/blob/6dd683f3e7a966bf0124444b746ab4a223bbe104/python/vyos/validate.py#L54-L56
I think it doesn't work with `netns`, so I added my own function

Comment (c-po) [reply python/vyos/ifconfig/interface.py]: Then please change the base function so we have only one instance of that function

File: python/vyos/ifconfig/interface.py (hunk 6 above)
Comment (sever-sever) [reply python/vyos/ifconfig/interface.py]: But this method will be per any interface change anyway. Even if we change the 'description,' isn't it?

File: python/vyos/ifconfig/interface.py (hunk 2 above)
Comment (sever-sever) [reply python/vyos/ifconfig/interface.py]: Done! Netns removed from def __init__ arg

File: python/vyos/ifconfig/interface.py (hunk 4 above)
Comment (sever-sever) [reply python/vyos/ifconfig/interface.py]: Done!

File: python/vyos/ifconfig/interface.py (hunk 5 above)
Comment (sever-sever) [reply python/vyos/ifconfig/interface.py:393]: For delete, we can't get data from `self.config{'netns']`
It will be empty config

File: python/vyos/ifconfig/interface.py (hunk 1 above)
Comment (sever-sever) [reply python/vyos/ifconfig/interface.py]: It should be in a separate commit, as required container build changes.
Due to `jmespath` not existing, it  cannot make the deb pkg
```

======================================================================
ERROR: test_is_addr_assigned (test_validate.TestVyOSValidate.test_is_addr_assigned)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vyos/work/T5241-control-edition/vyos-1x/src/tests/test_validate.py", line 25, in test_is_addr_assigned
    self.assertTrue(vyos.validate.is_addr_assigned('127.0.0.1'))
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/vyos/work/T5241-control-edition/vyos-1x/python/vyos/validate.py", line 87, in is_addr_assigned
    if is_intf_addr_assigned(interface, ip_address):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/vyos/work/T5241-control-edition/vyos-1x/python/vyos/validate.py", line 52, in is_intf_addr_assigned
    import jmespath
ModuleNotFoundError: No module named 'jmespath'

```

File: python/vyos/ifconfig/interface.py (hunk 3 above)
Comment (sever-sever) [reply python/vyos/ifconfig/interface.py]: Done!
```
    def exists(cls, ifname, netns=None) -> bool:
        cmd = f'ip link show dev {ifname}'
        if netns:
           cmd = f'ip netns exec {netns} {cmd}'
        return run(cmd) == 0

```

File: python/vyos/ifconfig/interface.py (hunk 5 above)
Comment (c-po) [reply python/vyos/ifconfig/interface.py:393]: Thanks

File: python/vyos/ifconfig/interface.py (hunk 6 above)
Comment (c-po) [reply python/vyos/ifconfig/interface.py]: You are right - that code-path could be triggered by the interface defaults

File: python/vyos/ifconfig/interface.py (hunk 1 above)
Comment (c-po) [reply python/vyos/ifconfig/interface.py]: Can you send the PR for container update?

Comment (sever-sever) [reply python/vyos/ifconfig/interface.py]: Sure https://github.com/vyos/vyos-build/pull/354

Comment (sever-sever) [reply python/vyos/ifconfig/interface.py]: > Then please change the base function so we have only one instance of that function

Done!

File: python/vyos/ifconfig/control.py [hunk 7]
```diff
@@ -49,6 +49,18 @@ def _popen(self, command):
         return popen(command, self.debug)
 
     def _cmd(self, command):
+        import re
+        if 'netns' in self.config:
+            # This command must be executed from default netns 'ip link set dev X netns X'
```
Comment (c-po) [inline python/vyos/ifconfig/control.py:54]: IMHO this is too much logic in two places.

looking at e.g.

```python
        elif is_intf_addr_assigned(self.ifname, addr, netns=self.config.get('netns')):
            exec_within_netns = f'ip netns exec {self.config.get("netns")}' if self.config.get('netns') else ''
            self._cmd(f'{exec_within_netns} ip addr del "{addr}" dev "{self.ifname}"')
```

you specify the netns command outside. In this helper you check if it was specified and execute it or add it if it needs to be added. Why not keep the existing code "as is" in the above case and only rely on you new logic here to "add" the netns command if required?

Thinking about execution time you always to regular matching on "every" command that traverses `self._cmd()` instead of simply extending a string. Can you please ensure that `ip netns exec` is only prepended once in this function and all other functions stay as they were?

That was one of the reasons to design a command wrapper so only single changes would be required when mangling the OS commands.

**Review by c-po — CHANGES_REQUESTED** [review]

File: python/vyos/ifconfig/control.py (hunk 7 above)
Comment (sever-sever) [reply python/vyos/ifconfig/control.py:54]: It doesn't work otherwise.
There are methods in the `interface.py` the method `ifconfig.interface.get_foo` will work fine with netns command which added a namespace from `control._cmd`
But if this method is used for example in another method, for example `set_foo` will be executed in netns (control._cmd), but if he called `get_foo` or external `ifconfig.validate.xxx` that method will be executed in the default namespace.

`get_foo` netns ok `control._cmd`
`set_foo` uses `get_foo` - **set_foo** will be executed in expected netns, the **get_foo** as I understand will be executed in default netns

`is_intf_addr_assigned` will be executed without nents even if we have `control._cmd`  the set the correct netns.
Yes, it is annoying. It looks ugly, maybe external lib [pyroute2](https://github.com/svinota/pyroute2) can help us with all netns magic

**Comment by frebib** [comment]
When moving an interface from one netns to the other, I guess it creates a new interface instead of moving it from the old netns to the new one?
```
vyos@vyos# set interfaces dummy dum0 netns ns02
[edit]
vyos@vyos# compare
[interfaces dummy dum0]
- netns "ns01"
+ netns "ns02"

[edit]
vyos@vyos# commit
[edit]
vyos@vyos# sudo ip netns exec ns02 ip l
1: lo: <LOOPBACK> mtu 65536 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
4: dum0: <BROADCAST,NOARP> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether 2e:f4:df:84:e9:9e brd ff:ff:ff:ff:ff:ff
[edit]
vyos@vyos# sudo ip netns exec ns01 ip l
1: lo: <LOOPBACK> mtu 65536 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
3: dum0: <BROADCAST,NOARP> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether f2:cd:b4:4d:ed:3b brd ff:ff:ff:ff:ff:ff
[edit]
vyos@vyos# show
 interfaces {
     dummy dum0 {
         address 1.2.3.4/32
         netns ns02
     }
     ethernet eth0 {
         hw-id 52:54:00:90:72:74
     }
     loopback lo {
     }
 }
 netns {
     name ns01 {
     }
     name ns02 {
     }
 }
```

**Comment by sever-sever** [comment]
> When moving an interface from one netns to the other, I guess it creates a new interface instead of moving it from the old netns to the new one?

@frebib Yes, I think it is due to this skip https://github.com/vyos/vyos-1x/pull/2022/files#diff-a4d8965b41a529a70fe69baf25f8413122f3b686ce0b9fe5f9033b4792f4d475R55-R64
Otherwise, we are in a situation where we set the interface in netns to the  same netns
```
            # exclude set netns cmd from netns to avoid:
            # failed to run command: ip netns exec ns01 ip link set dev veth20 netns ns01
```

There are also many other nuances/exceptions
